cql3/values.hh

```diff
@@ -30,179 +29,252 @@
-        if (auto pdata = std::get_if<fragmented_temporary_buffer::view>(&_data)) {
-            return *pdata;
-        }
-        return {};
+        return _data.index() <= 1;
     }
     explicit operator bool() const {
         return is_value();
     }
-    const fragmented_temporary_buffer::view* operator->() const {
-        return &std::get<fragmented_temporary_buffer::view>(_data);
-    }
-    const fragmented_temporary_buffer::view& operator*() const {
-        return std::get<fragmented_temporary_buffer::view>(_data);
-    }
-
-    bool operator==(const raw_value_view& other) const {
-        if (_data.index() != other._data.index()) {
-            return false;
-        }
-        if (is_value() && **this != *other) {
-            return false;
+    raw_value_view to_view() const;
+    bytes to_bytes() const {
+        if (auto p = std::get_if<bytes>(&_data)) {
+            return *p;
+        } else {
+            return ::to_bytes(std::get<managed_bytes>(_data));
         }
-        return true;
     }
-    bool operator!=(const raw_value_view& other) const {
-        return !(*this == other);
-    }
-
-    friend std::ostream& operator<<(std::ostream& os, const raw_value_view& value);
+    friend class raw_value_view;
 };
 
-/// \brief Raw CQL protocol value.
+/// \brief View to a raw CQL protocol value.
 ///
-/// The `raw_value` type represents an uninterpreted value from the CQL wire
-/// protocol. A raw value can hold either a null value, an unset value, or a byte
-/// blob that represents the value.
-class raw_value {
-    std::variant<bytes, null_value, unset_value> _data;
+/// \see raw_value
+class raw_value_view {
+public:
+    class view {
+    public:
+        using fragment_type = managed_bytes_view::fragment_type;
+    private:
+        // We use managed_bytes as a general-purpose fragmented buffer,
+        // so we want raw_value_view to handle managed_bytes_view,
+        // but we also want raw_value_view to handle fragmented_temporary_buffer::view without copying,
+        // because that's what's coming from the wire. Therefore, raw_value_view is a
+        // variant of managed_bytes_view and fragmented_temporary_buffer::view.
+        //
+        // Since the only part that differs between fragmented_temporary_buffer::view
+        // and managed_bytes_view is the method of getting the next fragment, which happens rarely,
+        // and the frequently used parts (accessing the first fragment and the total size)
+        // are common for both view types, we don't use a std::variant, and instead we store the
+        // common parts directly in raw_value_view, and only dispatch the action of getting the next
+        // fragment to the actual view type.
+        
+        using mb = managed_bytes_view::next_fragments_opaque;
+        using ftbv = fragmented_temporary_buffer::view::next_fragments_opaque;
+
+        std::variant<mb, ftbv, null_value, unset_value> _content;
```

Comment:
I see the variant remains.

Reply:
More than that, you have two separate branches for the two view types, so you're relying on the variant still.

You could remove the reliance on the variant by defining


```c++
struct fragmented_view_fragment_iterator {
    uintptr_t opaque;
    bytes_view (*next_fragment)(fragmented_view_fragment_iterator& fvfi);
};
```

These two pieces of data allow any fragmented view to work, provided state can be captured in a single uintptr_t. However, as long as the top-level variant exists, I don't see the advantage. The variant generates some bloat, yes, but it's not too terrible.

Reply:
The variant remains, but in this branch it is only ever used to cross the fragment boundary. Previously, the variant stored the whole views, as black boxes, now it only stores an opaque piece of data (the size of a pointer) needed to get the next fragment.

> You could remove the reliance on the variant by defining [...] These two pieces of data allow any fragmented view to work, provided state can be captured in a single uintptr_t.

That's pretty much what happens in this branch. Except the uintptr_t is typed and sits inside the variant, and instead of a function pointer, we have the index of the variant to decide what to call.
We don't want to eliminate the variant itself anyway, because it also handles null_value and unset_value, so putting the opaque piece of data inside the variant saves us 8 bytes for the function pointer. This can be easily changed any time to use a function pointer, since it won't affect the interface.

But more importantly, the interface of raw_value_view in this branch is different. In this branch, cql3::raw_value_view::view can be passed around as a generic FragmentedView, while in https://github.com/scylladb/scylla/pull/8160 we had to visit the variant with a lambda every time we want to look at the content (muddying the code and possibly causing some bloat).

That said, we don't do too much with this view, so perhaps adding a method for everything we do with it (deserialization, validation) to raw_value_view (this way we keep all the visiting in one place), like is done in 8160 is fine and I'm just badly overthinking it, especially since it doesn't seem like we look at the data inside raw_value_view multiple times.

But at the end of the day, I'm bringing this alternative up now, instead of just moving forward with the variant approach, because it's a matter of the interface.

Reply:
I see. So you retain a reference to the variant and all the demultiplexing is done in remove_view().

Probably there is no performance impact with the approach, and it does remove bloat.